core/trino-main/src/main/java/io/trino/operator/scalar/DataSizeFunctions.java

```diff
@@ -41,7 +57,39 @@ private DataSizeFunctions() {}
     public static Int128 parsePrestoDataSize(@SqlType("varchar(x)") Slice input)
     {
         String dataSize = input.toStringUtf8();
+        int valueLength = getValueLength(dataSize);
+        BigDecimal value = parseValue(dataSize.substring(0, valueLength), dataSize);
+        Unit unit = Unit.parse(dataSize.substring(valueLength), dataSize);
+        BigInteger bytes = value.multiply(unit.getFactor()).toBigInteger();
+        try {
+            return Decimals.valueOf(bytes);
+        }
+        catch (ArithmeticException e) {
+            throw new TrinoException(NUMERIC_VALUE_OUT_OF_RANGE, format("Value out of range: '%s' ('%sB')", dataSize, bytes));
+        }
+    }
+
+    @Description("Converts decimal data size string to bytes")
+    @ScalarFunction("parse_data_size_decimal")
+    @LiteralParameters("x")
+    @SqlType("decimal(38,0)")
+    public static Int128 parseDataSizeDecimal(@SqlType("varchar(x)") Slice input)
```

Comment:
What's the difference between this function and the one above?

Reply:
This one is using base 1000, the other one base 1024. Both are using SI units (kilo, mega), not IEC8000 units (kibi, mebi), to avoid any backward compatibility issues.